service/history/nDCBranchMgr.go

```diff
@@ -0,0 +1,184 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	ctx "context"
+	"time"
+
+	"github.com/pborman/uuid"
+	"github.com/uber/cadence/.gen/go/shared"
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/cluster"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/persistence"
+)
+
+type (
+	nDCBranchMgr interface {
+		prepareVersionHistory(
+			ctx ctx.Context,
+			incomingVersionHistory *persistence.VersionHistory,
+		) (versionHistoryIndex int, retError error)
+	}
+
+	nDCBranchMgrImpl struct {
+		context         workflowExecutionContext
+		mutableState    mutableState
+		shard           ShardContext
+		clusterMetadata cluster.Metadata
+		historyV2Mgr    persistence.HistoryV2Manager
+		logger          log.Logger
+	}
+)
+
+var _ nDCBranchMgr = (*nDCBranchMgrImpl)(nil)
+
+func newNDCBranchMgr(
+	context workflowExecutionContext,
+	mutableState mutableState,
+	shard ShardContext,
+	historyV2Mgr persistence.HistoryV2Manager,
+	logger log.Logger,
+) *nDCBranchMgrImpl {
+
+	return &nDCBranchMgrImpl{
+		context:         context,
+		mutableState:    mutableState,
+		shard:           shard,
+		clusterMetadata: shard.GetService().GetClusterMetadata(),
+		historyV2Mgr:    historyV2Mgr,
+		logger:          logger,
+	}
+}
+
+func (r *nDCBranchMgrImpl) prepareVersionHistory(
+	ctx ctx.Context,
+	incomingVersionHistory *persistence.VersionHistory,
+) (int, error) {
+
+	localVersionHistories := r.mutableState.GetVersionHistories()
+
+	versionHistoryIndex, lcaVersionHistoryItem, err := localVersionHistories.FindLCAVersionHistoryIndexAndItem(
+		incomingVersionHistory,
+	)
+	if err != nil {
+		return 0, err
+	}
+	versionHistory, err := localVersionHistories.GetVersionHistory(versionHistoryIndex)
+	if err != nil {
+		return 0, err
+	}
+
+	// if can directly append to a branch
+	if versionHistory.IsLCAAppendable(lcaVersionHistoryItem) {
+		return versionHistoryIndex, nil
+	}
+
+	newVersionHistory, err := versionHistory.DuplicateUntilLCAItem(lcaVersionHistoryItem)
+	if err != nil {
+		return 0, err
+	}
+
+	newVersionHistoryIndex, err := r.createNewBranch(
+		ctx,
+		versionHistory.GetBranchToken(),
+		lcaVersionHistoryItem.GetEventID(),
+		newVersionHistory,
+	)
+	if err != nil {
```

Comment:
just return r.createNewBranch(...)

Reply:
this is intentional, after the code is stabilized, errors will be logged here